pkg/controller/awscluster/awscluster_controller.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	// Always close the scope when exiting this function so we can persist any AWSMachine changes.
+	defer func() {
+		if err := scope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !awsCluster.DeletionTimestamp.IsZero() {
+		return reconcileDelete(scope)
+	}
+
+	// Handle non-deleted clusters
+	return reconcileNormal(scope)
+}
+
+// TODO(ncdc): should this be a function on scope?
+func reconcileDelete(scope *scope.ClusterScope) (reconcile.Result, error) {
+	scope.Info("Reconciling Cluster delete")
+
+	ec2svc := ec2.NewService(scope)
+	elbsvc := elb.NewService(scope)
+
+	if err := elbsvc.DeleteLoadbalancers(); err != nil {
+		scope.Error(err, "Error deleting cluster load balancer")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	if err := ec2svc.DeleteBastion(); err != nil {
+		scope.Error(err, "Error deleting cluster bastion")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	if err := ec2svc.DeleteNetwork(); err != nil {
+		scope.Error(err, "Error deleting cluster network")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// Cluster is deleted so remove the finalizer.
+	scope.AWSCluster.Finalizers = util.Filter(scope.AWSCluster.Finalizers, infrastructurev1alpha2.ClusterFinalizer)
+
+	return reconcile.Result{}, nil
+}
+
+// TODO(ncdc): should this be a function on scope?
+func reconcileNormal(scope *scope.ClusterScope) (reconcile.Result, error) {
+	scope.Info("Reconciling Cluster")
+
+	awsCluster := scope.AWSCluster
+
+	// If the AWSCluster doesn't have our finalizer, add it.
+	if !util.Contains(awsCluster.Finalizers, infrastructurev1alpha2.ClusterFinalizer) {
+		awsCluster.Finalizers = append(awsCluster.Finalizers, infrastructurev1alpha2.ClusterFinalizer)
+	}
+
+	ec2Service := ec2.NewService(scope)
+	elbService := elb.NewService(scope)
+
+	if err := ec2Service.ReconcileNetwork(); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to reconcile network for cluster %s/%s", awsCluster.Namespace, awsCluster.Name)
+	}
+
+	if err := ec2Service.ReconcileBastion(); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to reconcile bastion host for cluster %s/%s", awsCluster.Namespace, awsCluster.Name)
+	}
+
+	if err := elbService.ReconcileLoadbalancers(); err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to reconcile load balancers for cluster %s/%s", awsCluster.Namespace, awsCluster.Name)
+	}
+
+	if awsCluster.Status.Network.APIServerELB.DNSName == "" {
+		scope.Info("Waiting on API server ELB DNS name")
+		return reconcile.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	// Set APIEndpoints so the Cluster API Cluster Controller can pull them
+	awsCluster.Status.APIEndpoints = []infrastructurev1alpha2.APIEndpoint{
+		{
+			Host: awsCluster.Status.Network.APIServerELB.DNSName,
+			// TODO(ncdc): should this come from awsCluster.Status.Network.APIServerELB.Listeners[0].Port?
```

Comment:
If we can assume only a single listener, then I do not see why not.

Reply:
It looks like our ELB creation code sets up a single listener. Is it possible for a user to BYO ELB?

Reply:
I don't *think* so

Reply:
It looks like the code does get-or-create?